Katello/katello
# Fixes #34437 - Exclude filter can exclude excess errata

#### What are the changes introduced in this pull request?
This PR changes how errata RPMs get filtered for a content view version.  The main issue is that RPMs can be associated to multiple errata.  If one erratum excludes RPMs, then the errata that share those RPMs will also be excluded.  This is caused by the erratum RPM filtering code not knowing the complete list of errata that should be present in the destination repo after the copy call.  Instead of considering which RPMs to keep because they are part of an included erratum, the logic simply excludes all RPMs if that erratum is to be excluded.

This PR stops RPMs and module streams from being "added to the exclusion list" if other errata need them.

**There is also a change in behavior for copying filtered errata**.  Previously we always filtered out errata that had no packages and errata that didn't have any matching packages with the source repository. That means these errata will always be filtered out.  Errata should not be filtered out unless a user wants them to be, so I changed the logic to keep them.  I believe we always intended for this to be the case, but they were unintentionally filtered out.

#### Considerations taken when implementing this change?
This is a big change, so I tried to work within the existing code as much as possible.  During copying, I added a method `errata_exclusion_helpers` which returns two arrays: the list of errata to copy (`errata_to_ensure_complete`) and the complete list of errata that should be excluded (`all_excluded_errata`).  `errata_to_ensure_complete` is handed to the methods that calculate which RPMs and module streams should be added to the respective include and exclude filter lists.`all_excluded_errata` ensures that the errata that should be excluded are excluded at the end.  This second list is necessary because an erratum that should be excluded may be "completed" by the erratum RPM filtering code if other errata share all of its content.  In that case, the erratum would be included erroneously.

#### What are the testing steps for this pull request?
1) Sync the local `zoo` repository from the test_repos in any way you see fit
  - I added a new erratum `KATELLO-RHEA-2022:98765`.  It shares RPMs with `KATELLO-RHEA-2010:99143` and `KATELLO-RHEA-2010:0002`.
2) Add the repo to a content view and try excluding `KATELLO-RHEA-2022:98765`.  No other errata should get excluded and all of the RPMs should remain.
3) Sync the Red Hat repository with label `advanced-virt-for-rhel-8-x86_64-rpms`
4) Add it to a content view
5) Create an include filter and set the end date as 11 September 2021
6) Create an exclude filter and set the start date as 01 September 2020
7) Ensure the counts are the following:
```
Satellite 6.9:
988 Packages
16 Errata ( 5  11  0  )
16 Module Streams
```
8) Try all the above with dependency solving on
9) Test other content view filtering scenarios.



End notes:

I think it would be good to add some BATS tests to included this errata filtering scenario.  Content view filtering is difficult to test via unit tests.
